trace-normalization/src/normalize_utils.rs

```diff
@@ -25,23 +32,141 @@ pub(crate) fn truncate_utf8(s: &str, limit: usize) -> &str {
     s
 }
 
-// NormalizeService returns a span service or an error describing why normalization failed.
-// TODO: Implement this in a future PR
-// pub fn normalize_service(svc: String, lang: String) -> (String, Option<errors::NormalizeErrors>) {
-// if svc == "" {
-//     return (fallback_service(lang), errors::NormalizeErrors::ErrorEmpty);
-// }
-// if svc.len() > MAX_SERVICE_LEN {
-//     return (truncate_utf8(svc, MAX_SERVICE_LEN), errors::NormalizeErrors::ErrorTooLong.into());
-// }
-// TODO: implement tag normalization
-// let s: String = normalize_tag(svc);
-// if s == "" {
-//     return (fallbackService(lang), errors::NormalizeErrors::ErrorInvalid)
-// }
-// return (s, err)
-// (svc, None)
-// }
+// fallback_service returns the fallback service name for a service
+// belonging to language lang.
+pub(crate) fn fallback_service() -> String {
+    // In the go agent implementation, if a lang was specified in TagStats
+    // (extracted from the payload header) the fallback_service name would be "unnamed-{lang}-service".
+    DEFAULT_SERVICE_NAME.to_string()
+}
+
+// normalize_service normalizes a span service and returns an error describing the reason
+// (if any) why the name was modified.
+pub(crate) fn normalize_service(svc: &str) -> anyhow::Result<String> {
+    anyhow::ensure!(!svc.is_empty(), "Normalizer Error: Empty service name.");
+
+    let truncated_service = if svc.len() > MAX_SERVICE_LEN {
+        truncate_utf8(svc, MAX_SERVICE_LEN)
+    } else {
+        svc
+    };
+
+    normalize_tag(truncated_service)
+}
+
+// normalize_tag applies some normalization to ensure the tags match the backend requirements.
+pub(crate) fn normalize_tag(tag: &str) -> anyhow::Result<String> {
+    // Fast path: Check if the tag is valid and only contains ASCII characters,
+    // if yes return it as-is right away. For most use-cases this reduces CPU usage.
+    if is_normalized_ascii_tag(tag) {
+        return Ok(tag.to_string());
+    }
+
+    anyhow::ensure!(!tag.is_empty(), "Normalizer Error: Empty tag name.");
+
+    // given a dummy value
+    let mut last_char: char = 'a';
+
+    let mut result = String::with_capacity(tag.len());
+
+    let char_vec: Vec<char> = tag.chars().collect();
```

Comment:
I'm not sure you need to collect this into a vec, (which will allocate memory), since it's immutable you can just iterate over the original string:

```
for &char in tag.chars() {
```